app/models/csv_dumps/direct_to_object_storage.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module CsvDumps
+  # Custom 'Medium' duck type class to interface directly to
+  # a custom object storage location to securely host export files
+  # s3: (bucket, paths)
+  # Azure blob: (storage account, container, paths)
+  class DirectToObjectStorage
+    attr_reader :export_type, :adapter
+
+    class InsecureUploadDestination < StandardError; end
+
+    def initialize(export_type, adapter)
+      @export_type = export_type
+      @adapter = adapter
+    end
+
+    def metadata
+      {}
+    end
+
+    def save!
+      true
+    end
+
+    def put_file(gzip_file_path, compressed: true)
+      safe_for_private_upload?
+
+      storage_adapter.put_file(
+        object_storage_emails_export_path,
+        gzip_file_path,
+        put_file_opts
+      )
+    end
+
+    def storage_adapter
+      return @storage_adapter if @storage_adapter
+
+      storage_opts = {
+        # s3 adapter specific
+        bucket: ENV.fetch('EMAIL_EXPORT_S3_BUCKET', 'zooniverse-exports'),
+        prefix: ENV.fetch('EMAIL_EXPORT_S3_PREFIX', "#{Rails.env}/"),
+        # azure adapter specific
+        azure_storage_account: ENV.fetch('EMAIL_EXPORT_AZURE_STORAGE_ACCOUNT', 'zooniverse-exports'),
+        azure_storage_access_key: ENV['EMAIL_EXPORT_AZURE_STORAGE_ACCESS_KEY'],
+        azure_storage_container_public: ENV.fetch('EMAIL_EXPORT_AZURE_STORAGE_CONTAINER_PRIVATE', 'private'), # ensure private containers only
+        azure_storage_container_private: ENV.fetch('EMAIL_EXPORT_AZURE_STORAGE_CONTAINER_PRIVATE', 'private')
+      }
+
+      @storage_adapter = MediaStorage.send(:load_adapter, adapter, storage_opts)
+    end
+
+    private
+
+    def put_file_opts
+      {
+        private: true,
+        compressed: true,
+        content_disposition: "attachment; filename=\"#{export_file_name}.csv\""
+      }
+    end
+
+    def object_storage_emails_export_path(path='email_exports')
```

Comment:
I know that this was just modeled off of the `DirectToS3` code, but I'm wondering what the purpose is of constructing the path via `storage_adapter.stored_path`? The path for the email exports file is static, and lives in a separate bucket (and will live in a separate container) than other media uploads. Can't we construct the path without the use of the `stored_path` method? This just seems to be overcomplicating things in my mind, makes the code hard to read.

If we really do want to use the `stored_path` method, I would suggest potentially adding an optional `file_name` param, such that we could construct the `storage_path` in a single line like:
```ruby
storage_path = storage_adapter.stored_path('text/csv', path, export_file_name)
```

Reply:
Right - so instead i could just inline the modified functionality of `stored_path` method where needed. 

Perhaps something like this or similar?
``` ruby
# TODO figure out how to get the correct file extension from the mime type
# or just hard code it if we're only concerned with 'text/csv'
extension = extension_from_mime_type('text/csv')
"#{path}/#{export_file_name}.#{extension}"
```
private method to get the mime type
https://github.com/zooniverse/panoptes/blob/4cbea3e186595be93e2642dd0aa1442c37a8067d/app/services/media_storage/abstract_adapter.rb#L48